### ISSUE TITLE
call fflush(NULL) will get "not an sstable (bad magic number)"

### ISSUE THREAD

**Issue by egmkang** [google/leveldb#474]
it's easy to replay, you write  a program like this:
```
while (true)
    if RandomBetween(0, 10000) == 0
        fflush(NULL)
    leveldb::Set(RandomKey(), RandomValue())
```

you will get:
```
2017/06/01-15:24:37.659778 7fa9286dd700 Level-0 table #5975: 3703738 bytes Corruption: not an sstable (bad magic number)
```

**Comment by ghemawat** [comment]
Thanks for the minimal reproduction.  I was able to re-create the failure and I have a pretty strong hunch as to what is going wrong:

leveldb's file IO routines are using _unlocked routines (fread_unlocked, fwrite_unlocked, etc.)  We had assumed this was safe because our FILE* objects were only being used in a single-threaded context. However fflush(NULL) can come around the side and attempt to flush what we thought were internal private FILE objects, thereby causing a race.

I am working on a fix.

**Comment by Hypocritus** [comment]
@ghemawat, it's a pleasure to address you. Please correct me if I am wrong, but I believe that is bug is affecting many users. For instance, just the [STORJ](https://github.com/storj) project has many thousands of storage farmers (I being one), and I am regularly confronted with this error, which causes the node to abruptly exit.

[https://github.com/Storj/storjshare-daemon/issues/274](https://github.com/Storj/storjshare-daemon/issues/274)

**Comment by cmumford** [comment]
Fixed in 7e12c00ecf1bb725e212618e7026e4d34d6cd3bb
